magenta/models/polyphonic_rnn/polyphonic_rnn_lib.py

```diff
@@ -0,0 +1,1377 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for polyphonic RNN."""
+
+from collections import OrderedDict
+import copy
+import os
+import re
+import uuid
+
+# internal imports
+
+import numpy as np
+from scipy import linalg
+import tensorflow as tf
+
+import magenta.music as mm
+from magenta.protobuf import music_pb2
+
+
+##
+# begin metautils
+##
+
+
+def shape(x):
+  # Get shape of Variable through hacky hacky string parsing
+  shape_tup = repr(tf.Print(x, [tf.shape(x)])).split('shape=')[1]
+  shape_tup = shape_tup.split('dtype=')[0][:-1]
+  shape_tup = shape_tup.split(',')
+  shape_tup = [re.sub('[^0-9?]', '', s) for s in shape_tup]
+  # remove empty '' dims in 1D case
+  shape_tup = tuple([int(s) if s != '?' else -1
+                     for s in shape_tup if len(s) >= 1])
+  if sum([1 for s in shape_tup if s < 1]) > 1:
+    raise ValueError('too many ? dims')
+  return shape_tup
+
+
+def ndim(x):
+  return len(shape(x))
+
+
+# TODO: How can I correct this for shared / tied weights?
+def print_network(params):
+  n_params = sum([np.prod(shape(p)) for p in params])
```

Comment:
This probably isn't used in the end, I used it for debug mostly. Also it is currently not right (I don't think) given all the shared parameters in the final model